File: Design-Patterns-Using-Swift/Visitor/VisitorPattern.swift
protocol Element {
    func accept(visitor: Visitor)
}

class ConcreteElementA: Element {
    func accept(visitor: Visitor) {
        visitor.visitElementA(elementA: self)
    }
}

class ConcreteElementB: Element {
    func accept(visitor: Visitor) {
        visitor.visitElementB(elementB: self)
    }
}

protocol Visitor {
    func visitElementA(elementA: ConcreteElementA)
    func visitElementB(elementB: ConcreteElementB)
}

class ConcreteVisitor1: Visitor {
    func visitElementA(elementA: ConcreteElementA) {
        print("ConcreteVisitor1 visitElementA() method called.")
    }

    func visitElementB(elementB: ConcreteElementB) {
        print("ConcreteVisitor1 visitElementB() method called.")
    }
}

class ConcreteVisitor2: Visitor {
    func visitElementA(elementA: ConcreteElementA) {
        print("ConcreteVisitor2 visitElementA() method called.")
    }

    func visitElementB(elementB: ConcreteElementB) {
        print("ConcreteVisitor2 visitElementB() method called.")
    }
}

// Client code.
let visitor1 = ConcreteVisitor1()
let elementA = ConcreteElementA()
elementA.accept(visitor: visitor1)

let elementB = ConcreteElementB()
elementB.accept(visitor: visitor1)

/* 
ConcreteVisitor1 visitElementA() method called.
ConcreteVisitor1 visitElementB() method called.
*/

File: Design-Patterns-Using-Cpp/Visitor/VisitorPattern.cpp
#include <iostream>

class ConcreteElementA;
class ConcreteElementB;

// Visitor interface
class Visitor {
public:
    virtual void visitElementA(ConcreteElementA& elementA) = 0;
    virtual void visitElementB(ConcreteElementB& elementB) = 0;
    virtual ~Visitor() {}
};

// Element interface
class Element {
public:
    virtual void accept(Visitor& visitor) = 0;
    virtual ~Element() {}
};

// ConcreteElementA class
class ConcreteElementA : public Element {
public:
    void accept(Visitor& visitor) override {
        visitor.visitElementA(*this);
    }
};

// ConcreteElementB class
class ConcreteElementB : public Element {
public:
    void accept(Visitor& visitor) override {
        visitor.visitElementB(*this);
    }
};

// ConcreteVisitor1 class
class ConcreteVisitor1 : public Visitor {
public:
    void visitElementA(ConcreteElementA& elementA) override {
        std::cout << "ConcreteVisitor1 visitElementA() method called." << std::endl;
    }

    void visitElementB(ConcreteElementB& elementB) override {
        std::cout << "ConcreteVisitor1 visitElementB() method called." << std::endl;
    }
};

// ConcreteVisitor2 class
class ConcreteVisitor2 : public Visitor {
public:
    void visitElementA(ConcreteElementA& elementA) override {
        std::cout << "ConcreteVisitor2 visitElementA() method called." << std::endl;
    }

    void visitElementB(ConcreteElementB& elementB) override {
        std::cout << "ConcreteVisitor2 visitElementB() method called." << std::endl;
    }
};

// Client code
int main() {
    Visitor* visitor1 = new ConcreteVisitor1();
    Element* elementA = new ConcreteElementA();
    elementA->accept(*visitor1);

    Element* elementB = new ConcreteElementB();
    elementB->accept(*visitor1);

    delete visitor1;  // Freeing allocated memory
    delete elementA;  // Freeing allocated memory
    delete elementB;  // Freeing allocated memory

    return 0;
}

/*
ConcreteVisitor1 visitElementA() method called.
ConcreteVisitor1 visitElementB() method called.
*/